File: src/q3/q3.rs
const FILE: &str = "src/q3/input.txt";

#[allow(unused)]
pub mod part_1 {
    use std::collections::HashSet;
    use std::fs::read_to_string;

    use super::FILE;

    pub fn run() {
        match read_to_string(FILE).ok() {
            None => println!("could not read {FILE}"),
            Some(map) => println!("{}", solve(map.as_str())),
        }
    }

    fn solve(map: &str) -> u32 {
        let symbols = find_symbols(map);
        let numbers = find_numbers(map);

        numbers
            .iter()
            .filter(|(n, (row, start, end))| is_adjacdent(&symbols, *n, *row, *start, *end))
            .map(|(n, _)| n)
            .sum()
    }

    /**
     *
     * -----
     * -l  r
     * -----
     */
    fn is_adjacdent(
        symbols: &HashSet<String>,
        n: u32,
        row: usize,
        start: usize,
        end: usize,
    ) -> bool {
        // we must be very careful to not create a negative number from a usize

        // above
        // check cols start ... end to avoid overflow
        if row > 0 {
            for j in start..=end {
                if symbols.contains(&hash(row - 1, j)) {
                    return true;
                }
            }
        }
        // below
        for j in start..=end {
            if symbols.contains(&hash(row + 1, j)) {
                return true;
            }
        }

        // left column
        if start > 0 {
            if row > 0 {
                if symbols.contains(&hash(row - 1, start - 1)) {
                    return true;
                }
            }

            if symbols.contains(&hash(row, start - 1))
                || symbols.contains(&hash(row + 1, start - 1))
            {
                return true;
            }
        }

        // right column
        if row > 0 {
            if symbols.contains(&hash(row - 1, start + 1)) {
                return true;
            }
        }

        if symbols.contains(&hash(row, start + 1)) || symbols.contains(&hash(row + 1, start + 1)) {
            return true;
        }

        false
    }

    fn hash(row: usize, col: usize) -> String {
        format!("{row},{col}")
    }

    fn find_symbols(map: &str) -> HashSet<String> {
        map.lines()
            .enumerate()
            .flat_map(|(i, line)| {
                line.chars().enumerate().filter_map(move |(j, chr)| {
                    if chr.is_ascii_digit() || chr == '.' {
                        None
                    } else {
                        Some(hash(i, j))
                    }
                })
            })
            .collect::<HashSet<String>>()
    }

    /**
     * returns: [(number), (row, start, end)]
     */
    fn find_numbers(map: &str) -> Vec<(u32, (usize, usize, usize))> {
        let mut result = Vec::<(u32, (usize, usize, usize))>::new();
        for (i, line) in map.lines().enumerate() {
            // optionally store (start, digits in order)
            // the digits should be a u32 slice, but not sure how to unpack and concat these
            // reverse digits, then calculate powers in bases
            let mut last: Option<(usize, Vec<u32>)> = None;
            for (j, chr) in line.chars().enumerate() {
                match chr.to_digit(10) {
                    None => {
                        match last {
                            // end of a number
                            Some((old, digits)) => {
                                result.push((base_10_number(&digits), (i, old, j)));
                                last = None;
                            }
                            // continuation of no number
                            None => continue,
                        }
                    }
                    Some(n) => {
                        match last {
                            // start of a new number
                            None => last = Some((j, vec![n])),
                            // continuation of existing number
                            Some((old, mut digits)) => {
                                digits.push(n);
                                last = Some((old, digits));
                            }
                        }
                        // last digit means we should add on the number
                        if j == line.chars().count() - 1 {
                            if let Some((old, ref digits)) = last {
                                result.push((base_10_number(digits), (i, old, j + 1)));
                            }
                        }
                    }
                }
            }
        }
        result
    }

    fn base_10_number(digits: &[u32]) -> u32 {
        digits
            .iter()
            .rev()
            .enumerate()
            // cheat and use as
            .map(|(i, digit)| digit * u32::pow(10, i as u32))
            .sum()
    }

    #[test]
    fn test_base_10() {
        assert_eq!(base_10_number(&[1, 2, 3]), 123)
    }

    #[test]
    fn test_find_symbols() {
        let map = ".%.
..$";
        assert_eq!(
            find_symbols(map),
            HashSet::from(["0,1".to_string(), "1,2".to_string()])
        );
    }

    #[test]
    fn test_find_numbers_small() {
        let map = ".123..";
        assert_eq!(find_numbers(map), vec![(123, (0, 1, 4))])
    }

    #[test]
    fn test_find_numbers() {
        let map = ".123..
45.%6";

        assert_eq!(
            find_numbers(map),
            // TODO: fix indices
            vec![(123, (0, 1, 4)), (45, (1, 0, 2)), (6, (1, 4, 5))]
        )
    }
}

File: src/q3/q3.py
import re
from dataclasses import dataclass

import pytest

FILE = "input.txt"


def main():
    part_2()


@dataclass
class Number:
    row: int
    start: int  # inclusive
    end: int  # exlusice
    value: int


def part_1():
    with open(FILE, "r") as f:
        map = f.read().splitlines()

    numbers: list[Number] = []
    # hashes of coordinates
    symbols: set[str] = set()
    # preprocessing
    for row, line in enumerate(map):
        # numbers
        for number in re.finditer(r"[0-9]+", line):
            start, end = number.span()
            value = int(number.group(0))
            numbers.append(Number(row, start, end, value))

        # symbols
        for j, char in enumerate(line):
            if not char.isdigit() and char != '.':
                symbols.add(myhash(row, j))

    # compute answer
    result = 0
    for number in numbers:
        if is_adjacent(symbols, number):
            result += number.value

    print(f"answer : {result}")


def part_2():
    """map:
    "x,y" -> number,
    so
    123 -> {"0,0" : 0, "0,1" : 0, "0,2" : 0}
    too annoying to identify unique numbers, instead we can do this
    [123]
     0
    so we store indices of numbers, then maintain a unique set of these
    for all '*', check if there is an adjacdent number"""

    with open(FILE, "r") as f:
        result = solve_part_2(f.read().splitlines())
        print(f"answer : {result}")


def myhash(x: int, y: int) -> str:
    return f"{x},{y}"


def is_adjacent(symbols: set[str], number: Number) -> bool:
    for j in range(number.start - 1, number.end + 1):
        above = myhash(number.row - 1, j)
        below = myhash(number.row + 1, j)
        if above in symbols or below in symbols:
            return True

    left = myhash(number.row, number.start - 1)
    right = myhash(number.row, number.end)
    return left in symbols or right in symbols


def solve_part_2(map: list[str]) -> int:
    locations, numbers = preprocess(map)

    result = 0
    # asterisks
    for row, line in enumerate(map):
        for col, chr in enumerate(line):
            if chr == "*":
                adj = adjacdent_numbers(row, col, locations)
                if len(adj) == 2:
                    a, b = list(adj)
                    result += numbers[a] * numbers[b]

    return result


def preprocess(map: list[str]) -> tuple[dict[str, int], list[int]]:
    """Return locations, numbers"""
    locations: dict[str, int] = dict()
    numbers: list[int] = []

    # preprocessing
    for row, line in enumerate(map):
        # numbers
        for number in re.finditer(r"[0-9]+", line):
            start, end = number.span()
            value = int(number.group(0))
            for col in range(start, end):
                locations[myhash(col, row)] = len(numbers)
            numbers.append(value)

    return locations, numbers


def adjacdent_numbers(row: int, col: int, locations: dict[str, int]) -> set[int]:
    """Find indices of all adjacdent numbers for a given position"""
    result: set[int] = set()
    offsets: list[tuple[int, int]] = [
        # above
        (-1, -1), (0, -1), (1, -1),
        # left, right
        (-1, 0), (1, 0),
        # below
        (-1, 1), (0, 1), (1, 1)
    ]
    for offset in offsets:
        dx, dy = offset
        pos = myhash(col + dx, row + dy)
        if pos in locations:
            result.add(locations[pos])

    return result


def test_adjacdent_numbers():
    indices = adjacdent_numbers(1, 1, {
        "0,0": 0, "1,0": 1, "2,0": 2,
        "0,1": 3, "2,1": 4,
        "0,2": 5, "1,2": 6, "2,2": 7
    })
    assert indices == set([i for i in range(8)])


def test_part_2_small():
    map = ["467..114..",
           "...*......",
           "..35..633.",
           "......#...",
           "617*......",
           ".....+.58.",
           "..592.....",
           "......755.",
           "...$.*....",
           ".664.598.."]
    locations, _ = preprocess(map)
    assert (map[1][3] == "*")
    assert adjacdent_numbers(1, 3, locations) == set([0, 2])
    assert solve_part_2(map) == 467835


if __name__ == "__main__":
    main()
